src/core/thread/mle.cpp

```diff
@@ -1752,8 +1762,19 @@ ThreadError Mle::HandleDataResponse(const Message &aMessage, const Ip6::MessageI
                                 (mDeviceMode & ModeTlv::kModeFullNetworkData) == 0,
                                 networkData.GetNetworkData(), networkData.GetLength());
 
+    diff = 0;
+
 exit:
     (void)aMessageInfo;
+
+    if (diff > 0)
```

Comment:
In general, we should only send a Data Request message if we still need to update the Active and/or Pending Datasets.  I think it would be better to do as follows: if the Data Response does not include all of the information needed (i.e. network data tlv, active dataset, pending dataset) then no change to local state should occur and just send a Data Request message as below.  Does that make sense?